File: Maximum Number of coins.java
class Solution{
    int maxCoins(int N, ArrayList<Integer> arr) {

    // add 1 at the beginning and end of the list
        arr.add(0, 1);
        arr.add(1);
        
        int n = arr.size();
        int[][] dp = new int[n][n];
        
        // l is the length of the subarray
        for (int l = 3; l <= n; l++) {
            // i is the starting index of the subarray
            for (int i = 0; i <= n - l; i++) {
                // j is the ending index of the subarray
                int j = i + l - 1;
                // k is the index of the last balloon to be burst
                for (int k = i + 1; k < j; k++) {
                    dp[i][j] = Math.max(dp[i][j], dp[i][k] + dp[k][j] + arr.get(i) * arr.get(k) * arr.get(j));
                }
            }
        }
        
        return dp[0][n-1];
    }
    
}

 

File: Maximum Number of coins.cpp
class Solution{
    public:
        int t[405][405];
        int maxt(vector<int>&a,int i,int j){
            if(i>j)
            {
                return 0;
            }
            int ans=INT_MIN;
            if(t[i][j]!=-1)
            {
            return t[i][j];
            }
            for(int k=i;k<=j;k++){
                ans=max(ans,maxt(a,i,k-1)+maxt(a,k+1,j)+a[k]*(i-1>=0?a[i-1]:1)*(j+1<a.size()?a[j+1]:1));
            }
            return t[i][j]=ans;
        }
    int maxCoins(int N, vector<int> arr) {
        memset(t,-1,sizeof(t));
        int i=0;
        int j=arr.size()-1;
        return maxt(arr,i,j);
        //Write your code here
    }
};
